waltz-tools/src/main/java/com/wepay/waltz/tools/cluster/ClusterCli.java

```diff
@@ -364,6 +378,103 @@ private void buildStoragePartitionValidationResult(int numPartitions,
             }
         }
 
+        /**
+         * Builds REPLICA_RECOVERY_STATUS {@code PartitionValidationResults} for all the replicas in the
+         * cluster.
+         * The validation result is set to success if and only if:
+         * 1. The set of replicas in replica assignments is the same as the set of replicas in replica states.
+         * 2. For each replica, the session id in the replica states equals the session id in partitionMetadata,
+         * and closingHighWaterMark is UNRESOLVED.
+         *
+         * @param numPartitions Total number of partitions in the cluster.
+         * @param partitionRoot ZNode used to fetch replica states for each partition.
+         * @param replicaAssignments Replica assignments for all the partitions.
+         * @param zkClient ZooKeeper Client used to fetch partition metadata.
+         * @param partitionsValidationResultsList a {@code List<PartitionValidationResults>} that contains all the
+         *                                       validation results.
+         */
+        private void buildStorageRecoveryCompleteValidationResult(int numPartitions,
+                                                                  ZNode partitionRoot,
+                                                                  ReplicaAssignments replicaAssignments,
+                                                                  ZooKeeperClient zkClient,
+                                                                  List<PartitionValidationResults> partitionsValidationResultsList) {
+            // A set that contains all replica ids in replica assignments
+            Set<ReplicaId> assignmentReplicaIdSet = new TreeSet<>();
+            for (Map.Entry<String, int[]> entry : replicaAssignments.replicas.entrySet()) {
+                for (int partitionId : entry.getValue()) {
+                    assignmentReplicaIdSet.add(new ReplicaId(partitionId, entry.getKey()));
+                }
+            }
+
+            for (int partitionId = 0; partitionId < numPartitions; partitionId++) {
+                String error = "";
+                ZNode zNode = new ZNode(partitionRoot, Integer.toString(partitionId));
+                Map<ReplicaId, ReplicaState> replicaState;
+                Long sessionId;
+
+                try {
+                    PartitionMetadata partitionMetadata = zkClient.getData(zNode, PartitionMetadataSerializer.INSTANCE).value;
+                    replicaState = partitionMetadata.replicaStates;
+                    sessionId = partitionMetadata.sessionId;
+                } catch (KeeperException | ZooKeeperClientException e) {
+                    replicaState = Collections.emptyMap();
+                    sessionId = null;
+                    error = "Cannot obtain replica states";
+                }
+
+                Map<ReplicaId, ReplicaState> sortedReplicaState = new TreeMap<>(replicaState);
```

Comment:
any particular reason you are creating a new `sortedReplicaState`?

Reply:
The only purpose here is to keep the error msg ordered by ReplicaId when they are being added.

Reply:
Also suggested by @abhipardhu on this. Update it and remove it since ordering by `storageNodeConnectString` won't improve the readability of error msg, and msgs are already grouped by partitionId.

Reply:
Ya, order shouldn't matter.